rustls/src/client/hs.rs

```diff
@@ -289,16 +305,45 @@ fn emit_client_hello_for_retry(
     // We don't do renegotiation at all, in fact.
     cipher_suites.push(CipherSuite::TLS_EMPTY_RENEGOTIATION_INFO_SCSV);
 
+    let mut chp_payload = ClientHelloPayload {
+        client_version: ProtocolVersion::TLSv1_2,
+        random: input.random,
+        session_id: input.session_id,
+        cipher_suites,
+        compression_methods: vec![Compression::Null],
+        extensions: exts,
+    };
+
+    match (ech_context.as_mut(), config.grease_ech_hpke_provider) {
+        // ECH config, GREASE is irrelevant.
+        (Some(ech_context), _) => {
+            let ech_chp_payload = ech_context.ech_hello(chp_payload, retryreq)?;
+            chp_payload = ech_chp_payload;
+        }
+        // No ECH config, GREASE provider is present.
+        (None, Some(hpke_provider)) => {
+            let grease_ext = EchState::grease_ech_ext(
+                hpke_provider,
+                config.provider.secure_random,
+                input.server_name.clone(),
+                &chp_payload,
+            )?;
+            chp_payload.extensions.push(grease_ext);
+        }
+        // No ECH config, no GREASE ECH HPKE provider
+        _ => {}
+    }
+
+    // Note what extensions we sent.
+    input.hello.sent_extensions = chp_payload
+        .extensions
+        .iter()
+        .map(ClientExtension::get_type)
+        .collect();
+
     let mut chp = HandshakeMessagePayload {
         typ: HandshakeType::ClientHello,
-        payload: HandshakePayload::ClientHello(ClientHelloPayload {
-            client_version: ProtocolVersion::TLSv1_2,
-            random: input.random,
-            session_id: input.session_id,
-            cipher_suites,
-            compression_methods: vec![Compression::Null],
-            extensions: exts,
-        }),
+        payload: HandshakePayload::ClientHello(chp_payload),
```

Comment:
Nit: maybe do a separate commit that reorders the `ClientHelloPayload` instantiation and `sent_extensions` before adding the ECH stuff?

Reply:
Done.